src/opendkim.cc

```diff
@@ -1139,6 +1142,137 @@ const char *OpenDKIM::VerifyBase(OpenDKIM *obj, char *id)
+    return;
+  }
+
+  if (!info[0]->IsObject()) {
+    Nan::ThrowTypeError("diffheaders(): Argument should be an object");
+    return;
+  }
+  
+  maxcost = _value_to_int(info[0], "maxcost");
+
+  if(maxcost == 0 ){
+    Nan::ThrowTypeError("diffheaders(): Invalid maxcost cannot be 0");
+    return;
+  }
+  Isolate* isolate = info.GetIsolate(); // Needed to make new array
+  Local<Array> out_array = Array::New(isolate);
+
+  if (obj->dkim == NULL)
+  {
+    Nan::ThrowTypeError(
+        "diffheaders(): library must be initialized first");
+    return;
+  }
+
+  obj->GetSignature(info);
+
+  if (obj->sig == NULL)
+  {
+    Nan::ThrowTypeError(
+        "diffheaders():  either there was no signature or called before eom() or chunk_end()");
+    return;
+  }
+
+  statp = dkim_sig_getcanons(obj->sig, &hdrcanon, NULL);
+  
+  if (statp != DKIM_STAT_OK)
+  {
+    result = get_error(statp);
+    return;
+  }
+
+  if (&hdrcanon == NULL)
+  {
+    Nan::ThrowTypeError(
+        "diffheaders():  signature as no  header canonicalization mode.");
+    return;
+  }
+  
+  statp = dkim_ohdrs(obj->dkim, obj->sig, ohdrs, &nohdrs);
+  if (statp != DKIM_STAT_OK)
+  {
+    result = get_error(statp);
+    Nan::ThrowTypeError(result);
+    return;
+  }
+
+   if (ohdrs == NULL)
+  {
+    Nan::ThrowTypeError("diffheaders(): ohdrs is NULL");
+    return;
+  }
+  
+  statp = dkim_diffheaders(dkim, hdrcanon, maxcost, (char**) ohdrs, nohdrs,  out, &nout);
+
+  if (statp != DKIM_STAT_OK)
+  {
+    
+    if (statp == DKIM_STAT_NOTIMPLEMENT)
+    {
+      result = "Not implemented: The library must be compiled with an approximate regular\n expression library in order to provide this service.\n Check libopendkim is compiled with --with-tre optional packages";
+    }else{
+      result = get_error(statp);
+    }
+    Nan::ThrowTypeError(result);
+    return;
+  }
+
+  if (nout <= MAXHDRCOUNT)
+  {
+
+    v8::Local<v8::String> hd_old_key = Nan::New<v8::String>("hd_old").ToLocalChecked();
+    v8::Local<v8::String> hd_new_key = Nan::New<v8::String>("hd_new").ToLocalChecked();
+  
+    for (int i = 0; i < nout; i++)
+    {
+      // createobject holding  hd_old  hd_new
+      v8::Local<v8::Object> diffObject = Nan::New<v8::Object>();
+      v8::Local<v8::String> hd_old;
+      v8::Local<v8::String> hd_new;
+
+      if( (*out)[i].hd_old != NULL){
+        hd_old = Nan::New<v8::String>((char *) (*out)[i].hd_old).ToLocalChecked();
+      }else{
+        hd_old = Nan::New<v8::String>("NULL").ToLocalChecked();
+      }
+      if((*out)[i].hd_new != NULL){
+        hd_new = Nan::New<v8::String>((char *)(*out)[i].hd_new).ToLocalChecked();
+      }else{
+        hd_new = Nan::New<v8::String>("NULL").ToLocalChecked();
+      }
+      Nan::Set(diffObject, hd_old_key, hd_old);
+      Nan::Set(diffObject, hd_new_key, hd_new);
+      Nan::Set(out_array, i, diffObject);
+    }
+  } else {
+    Nan::ThrowTypeError("diffheaders(): too many headers to fit");
+    return;
```

Comment:
If we hit this case after building an elaborate `out_array` with tons of `diffObjects`, are we going to leak all that memory?  I honestly can't recall if this will get reclaimed or not.  I would hope that when we return it, the entire structure ends up with a refcount indicating it's in use, and if we just simply throw, then this never happens and the memory is eventually reclaimed.

Reply:
Since the out_array is declared as local and depends on an isolate  so when out of scope memory is reclaimed. In this case at the end of the function.
Look at this:
> Handles and garbage collection
> A handle provides a reference to a JavaScript object’s location in the heap. The V8 garbage collector reclaims memory used by objects that can no longer again be accessed. During the garbage collection process the garbage collector often moves objects to different locations in the heap. When the garbage collector moves an object the garbage collector also updates all handles that refer to the object with the object’s new location.
> 
> An object is considered garbage if it is inaccessible from JavaScript and there are no handles that refer to it. From time to time the garbage collector removes all objects considered to be garbage. V8’s garbage collection mechanism is key to V8’s performance.
> 
> There are several types of handles:
> 
> Local handles are held on a stack and are deleted when the appropriate destructor is called. These handles’ lifetime is determined by a handle scope, which is often created at the beginning of a function call. When the handle scope is deleted, the garbage collector is free to deallocate those objects previously referenced by handles in the handle scope, provided they are no longer accessible from JavaScript or other handles. This type of handle is used in the hello world example above.
> 
> Local handles have the class Local<SomeType>.
> 
> Note: The handle stack is not part of the C++ call stack, but the handle scopes are embedded in the C++ stack. Handle scopes can only be stack-allocated, not allocated with new.

from [https://v8.dev/docs/embed](https://v8.dev/docs/embed) guide.

From the c point of view as dkim_diffheader** out is a function local variable, when function is done then memory should be available. To be sure we could add a `free( (*out));`.